Add additional template functions

stuggi opened the pull request:
Adds templat functions which:
- to exec an inline template
- indent a template
- remove n continuous empty lines from a template

An example would be a template like:
```
{{define "my-template"}}new template

content

with empty lines

to
remove
{{end}}
Some other template content and add the rendered from my-template
{{$var := execTempl "my-template" . | removeNewLines 1}}
{{$var}}
```

stuggi commented:
an example for using it in nova for the nova.conf is in https://github.com/openstack-k8s-operators/nova-operator/pull/874

stuggi commented on .pre-commit-config.yaml:
```diff
@@ -40,4 +40,8 @@ repos:
+      exclude: |
+        (?x)(
+            ^vendor|
+            ^modules/common/util/template_util_test.go
```
I had to add this because of the test data for `TestRemoveNewLines` where I test preceding spaces

gibizer replied:
this feels a bit scary. I would the test data to normal strings with \n and \t to avoid this. But that can be done in a follow up.

stuggi replied:
I have removed the spaces from the test data and returned the config.

SeanMooney commented:
we coudl do this but i guess we were trying to avoid defienign our own dsl
I'm concurred that if we do this this way we may end up building a complex templaating system
so we should be very intentional about taking this direction if we do.

stuggi commented:
> we coudl do this but i guess we were trying to avoid defienign our own dsl I'm concurred that if we do this this way we may end up building a complex templaating system so we should be very intentional about taking this direction if we do.

yes we should not introduce a lot of/further complexity. At least right now I am not aware of something else needed. I can remove the `indent` func as it was not requested. What I like with this approach is that it uses the default way on how you  extend the go templating using these funcs.

SeanMooney commented:
> > we coudl do this but i guess we were trying to avoid defienign our own dsl I'm concurred that if we do this this way we may end up building a complex templaating system so we should be very intentional about taking this direction if we do.
> 
> yes we should not introduce a lot of/further complexity. At least right now I am not aware of something else needed. I can remove the `indent` func as it was not requested. What I like with this approach is that it uses the default way on how you extend the go templating using these funcs.

yep that is a plus as it not a net new way of doing this.
and this filter/fucntion style approach should feel famialr to anyone who has used jinja as well.

auniyal61 commented on modules/common/util/template_util.go:
```diff
@@ -138,6 +139,57 @@ func ExecuteTemplate(templateFile string, data interface{}) (string, error) {
+}
+
+// template function to remove empty lines if there are > n continuous empty lines
+func removeNewLines(n int, in string) (out string) {
```
IIUTC, this allow people to have n number of lines in between.
I think we may not need this in case conf files, but may be for other templates!

auniyal61 commented:
I think its better approach, as there are only template level changes at operator side.

right now, the removeNewLines function is not working as we want/expected, I had tested it locally.

[current_nova.conf render](https://privatebin.corp.redhat.com/?87d05bb8bff66009#2wjY8nDRDC4tRNsTWBQTmcoQffVReFxzrBQPHusQGCcv)

[removeNewLines(0, str)](https://privatebin.corp.redhat.com/?8ad6a5ecb2989b3d#5d1KP2LvAjr66ZpEeTfqFf5yApZhKK4vLqtP3g1n9EMm)

[removeNewLines(1, str)
](https://privatebin.corp.redhat.com/?5e35003c489a5510#GjsXjKHT1WFhiu5XhDjJkCw1BnDEYEuggXGtvAdSBjUg)

[removeNewLines(2, str)](https://privatebin.corp.redhat.com/?6bf8a9871eb0b92a#CrXiairiJJJBUn69R3KyjibzaAKYrXEecX6A7Ca3rZwE)

[removeNewLines(3, str)](https://privatebin.corp.redhat.com/?c9a0b9c30e054542#EmtG75reJBWHQAM4RFMdT9yx98pnDRwTywUe5y3oqSjb)


(but that may not be the focus of this patch)

stuggi commented:
> I think its better approach, as there are only template level changes at operator side.
> 
> right now, the removeNewLines function is not working as we want/expected, I had tested it locally.
> 
> [current_nova.conf render](https://privatebin.corp.redhat.com/?87d05bb8bff66009#2wjY8nDRDC4tRNsTWBQTmcoQffVReFxzrBQPHusQGCcv)
> 
> [removeNewLines(0, str)](https://privatebin.corp.redhat.com/?8ad6a5ecb2989b3d#5d1KP2LvAjr66ZpEeTfqFf5yApZhKK4vLqtP3g1n9EMm)
> 
> [removeNewLines(1, str) ](https://privatebin.corp.redhat.com/?5e35003c489a5510#GjsXjKHT1WFhiu5XhDjJkCw1BnDEYEuggXGtvAdSBjUg)


Isnt' this what you wanted? it will remove multiple continuous empty lines and just keep a single?
Not covering the special header case

> 
> [removeNewLines(2, str)](https://privatebin.corp.redhat.com/?6bf8a9871eb0b92a#CrXiairiJJJBUn69R3KyjibzaAKYrXEecX6A7Ca3rZwE)
> 
> [removeNewLines(3, str)](https://privatebin.corp.redhat.com/?c9a0b9c30e054542#EmtG75reJBWHQAM4RFMdT9yx98pnDRwTywUe5y3oqSjb)
> 
> (but that may not be the focus of this patch)

auniyal61 commented:
> > I think its better approach, as there are only template level changes at operator side.
> > right now, the removeNewLines function is not working as we want/expected, I had tested it locally.
> > [current_nova.conf render](https://privatebin.corp.redhat.com/?87d05bb8bff66009#2wjY8nDRDC4tRNsTWBQTmcoQffVReFxzrBQPHusQGCcv)
> > [removeNewLines(0, str)](https://privatebin.corp.redhat.com/?8ad6a5ecb2989b3d#5d1KP2LvAjr66ZpEeTfqFf5yApZhKK4vLqtP3g1n9EMm)
> 
> Isnt' this what you wanted? it will remove multiple continuous empty lines and just keep a single? Not covering the special header case
> 



yes, after adding for section-header with n =0, it should work
we need [this](https://logserver.rdoproject.org/57/857/0b3a0d59ae6b1ae5c0bba96da969a76b5fabd90b/github-check/nova-operator-tempest-multinode/0cbc344/controller/ci-framework-data/logs/openstack-k8s-operators-openstack-must-gather/namespaces/openstack/secrets/nova/nova-api-config-data.yaml-01-nova.conf).

stuggi commented:
> > > I think its better approach, as there are only template level changes at operator side.
> > > right now, the removeNewLines function is not working as we want/expected, I had tested it locally.
> > > [current_nova.conf render](https://privatebin.corp.redhat.com/?87d05bb8bff66009#2wjY8nDRDC4tRNsTWBQTmcoQffVReFxzrBQPHusQGCcv)
> > > [removeNewLines(0, str)](https://privatebin.corp.redhat.com/?8ad6a5ecb2989b3d#5d1KP2LvAjr66ZpEeTfqFf5yApZhKK4vLqtP3g1n9EMm)
> > 
> > 
> > Isnt' this what you wanted? it will remove multiple continuous empty lines and just keep a single? Not covering the special header case
> 
> yes, after adding for section-header with n =0, it should work we need [this](https://logserver.rdoproject.org/57/857/0b3a0d59ae6b1ae5c0bba96da969a76b5fabd90b/github-check/nova-operator-tempest-multinode/0cbc344/controller/ci-framework-data/logs/openstack-k8s-operators-openstack-must-gather/namespaces/openstack/secrets/nova/nova-api-config-data.yaml-01-nova.conf).

I have added a form of your cleanup func from https://github.com/openstack-k8s-operators/lib-common/pull/557 and added you as co-author to the commit msg. I was just not sure if you want your full name listed there as it is not shown in your github account. I can change that, just let me know.

SeanMooney commented on modules/common/util/template_util.go:
```diff
@@ -153,11 +230,16 @@ func lower(s string) string {
+		"indent":                   indent,
+		"lower":                    lower,
+		"removeNewLines":           removeNewLines,
+		"removeNewLinesInSections": removeNewLinesInSections,
```
this is proably the only concern i have with this approach.

this is a nice approach however if we want to go this route
then i think we should also allow passing extra functions via the template class

basically if we go this route i think it would make esne to enable the operator to populate template.ExtaFuncMap with functions defined in the operator.

that just avoid having to always update libcommon first and bumping the libcommon version for functions that may only make sense for a single operator.

that could be a separate PR, to make it possible for client of libcommon to pass in a function map that will be combined with the default ones provided by libcommon.

stuggi replied:
yes we can enhance this and could be used then from within the operators. but if people see the need to have additional functions, we should move them into here and can be re-used by others and not end up each of the operators create its own functions. but for specific use cases it makes sense. Lets do this in a follow up

stuggi replied:
I have created https://issues.redhat.com/browse/OSPRH-10672 to track it

SeanMooney replied:
ack sound good to me

SeanMooney commented:
reflecting on this over the weekend i think I'm satisfied that this is a better approach to achieve this usecase, i left one comment inline but I'm mostly happy with this as proposed.

auniyal61 commented:
lgtm
thanks for including unit tests and bar.conf as well

SeanMooney reviewed (approved):
this looks ok to proceed with from my perspective

auniyal61 commented:
secrets from nova-operator gate job
https://logserver.rdoproject.org/74/874/da8db1ca9e6535b95fd9567fc99a477fb8770b01/github-check/nova-operator-tempest-multinode/265ba27/controller/ci-framework-data/logs/openstack-k8s-operators-openstack-must-gather/namespaces/openstack/secrets/nova/nova-cell0-conductor-config-data.yaml-01-nova.conf

abays reviewed (approved):
/lgtm
